### PR TITLE
New resource: azurerm_purview_account

### LINKED ISSUE #10358: Support for Azure Purview
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I would like to provision azure purview accounts with terraform
https://azure.microsoft.com/en-us/services/purview/

### New or Affected Resource(s)

* azurerm_purview_account

### Potential Terraform Configuration

```hcl
resource "azurerm_purview_account" "example" {
  name                = "string"
  location            = "string"
  resource_group_name  = "string"
}
```

### References

* #0000

### PR DESCRIPTION
**Pull request by allantargino**
fixes #10358 
depends on #10393

### CODE REVIEW

**Comment by allantargino** [comment]
@katbyte @tombuildsstuff I would appreciate whenever you are free to do a review on this PR 😄 Thanks!

**Comment by allantargino** [comment]
Hey @katbyte @tombuildsstuff @mbfrahry, would you please take a look on this PR? Thank you!

**Review by favoretti — CHANGES_REQUESTED** [review]
Hi @allantargino, thank you for this PR. Looking mostly good to me, if we could address some minor issues, we should be able to get this over the finish line.

File: azurerm/internal/services/purview/purview_account_resource.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_purview_account", *existing.ID)
+		}
+	}
+
+	account := purview.Account{
+		AccountProperties: &purview.AccountProperties{},
+		Identity: &purview.Identity{
+			Type: purview.SystemAssigned,
```
Comment (katbyte) [inline azurerm/internal/services/purview/purview_account_resource.go:147]: is this the only value we can assign here?

File: azurerm/internal/services/purview/purview_account_resource.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+		},
+		Location: &location,
+		Sku: &purview.AccountSku{
+			Capacity: utils.Int32(int32(d.Get("sku_capacity").(int))),
+			Name:     purview.Standard,
```
Comment (katbyte) [inline azurerm/internal/services/purview/purview_account_resource.go:152]: is there a reason we are not exposing the sku as a property?

File: website/docs/r/purview_account.html.markdown [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+The following arguments are supported:
+
+* `location` - (Required) The Azure Region where the Purview Account should exist. Changing this forces a new Purview Account to be created.
+
+* `name` - (Required) The name which should be used for this Purview Account. Changing this forces a new Purview Account to be created.
+
+* `resource_group_name` - (Required) The name of the Resource Group where the Purview Account should exist. Changing this forces a new Purview Account to be created.
+
+* `sku_capacity` - (Required) SKU capacity for platform size and catalog capabilities. Accepted values are `4` and `16`.
```
Comment (katbyte) [inline website/docs/r/purview_account.html.markdown:39]: grammer
```suggestion
* `sku_capacity` - (Required) The SKU's capacity for platform size and catalog capabilities. Accepted values are `4` and `16`.
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @allantargino - overall this looks good, i've left some comments inline that once addressed this should be good to merge

**Comment by allantargino** [comment]
Thanks @katbyte, I just iterated over your comments :)

File: azurerm/internal/services/purview/purview_account_resource.go (hunk 2 above)
Comment (favoretti) [reply azurerm/internal/services/purview/purview_account_resource.go:152]: Dunno if I'm off-base here, but I assume they will introduce more, can we expose complete property with the only accepted value of `Standard` for the name? Would save time when it's exposed later on for the future generations wanting to expand this one?

Comment (allantargino) [reply azurerm/internal/services/purview/purview_account_resource.go:152]: That makes sense. I will expose this property.

File: azurerm/internal/services/purview/purview_account_resource.go [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+			"sku_capacity": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ValidateFunc: validation.IntInSlice([]int{
+					4,
+					16,
+				}),
+			},
+
+			"sku_name": {
```
Comment (favoretti) [inline azurerm/internal/services/purview/purview_account_resource.go:61]: @allantargino Sorry for being unclear there, I’ve been chatting to @katbyte a few weeks back on how we define SKU, basically I believe we want `sku_name` to contain both name and capacity and parse them inline, i.e. `sku_name` property will allow for `Standard_4` and `Standard_16` as values.

@katbyte that was the conclusion, right?

Comment (katbyte) [reply azurerm/internal/services/purview/purview_account_resource.go:61]: yes typically thats how we've been handling the sku names as of late

Comment (allantargino) [reply azurerm/internal/services/purview/purview_account_resource.go:61]: Oh I see, let me refactor that then. Thanks for the inputs :)

**Review by katbyte — APPROVED** [review]
Thanks @allantargino - this LGTM 👍

File: azurerm/internal/services/purview/purview_account_resource.go [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+		existing, err := client.Get(ctx, resourceGroup, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("Error checking for presence of existing Purview Account %q (Resource Group %q): %s", name, resourceGroup, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_purview_account", *existing.ID)
+		}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/purview/purview_account_resource.go:141]: can we use the Terraform managed resource ID here, making this:

```suggestion
		if !utils.ResponseWasNotFound(existing.Response) {
			return tf.ImportAsExistsError("azurerm_purview_account", id.ID())
		}
```

File: azurerm/internal/services/purview/purview_account_resource_test.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+				check.That(data.ResourceName).Key("sku_name").HasValue("Standard_4"),
+				check.That(data.ResourceName).Key("public_network_enabled").HasValue("true"),
+				check.That(data.ResourceName).Key("identity.0.type").Exists(),
+				check.That(data.ResourceName).Key("identity.0.principal_id").Exists(),
+				check.That(data.ResourceName).Key("identity.0.tenant_id").Exists(),
+				check.That(data.ResourceName).Key("catalog_endpoint").Exists(),
+				check.That(data.ResourceName).Key("guardian_endpoint").Exists(),
+				check.That(data.ResourceName).Key("scan_endpoint").Exists(),
+				check.That(data.ResourceName).Key("atlas_kafka_endpoint_primary_connection_string").Exists(),
+				check.That(data.ResourceName).Key("atlas_kafka_endpoint_secondary_connection_string").Exists(),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/purview/purview_account_resource_test.go:37]: we can remove all of these in favour of an "import" step which validates the remote fields match the fields in the config:

```suggestion

```

File: azurerm/internal/services/purview/purview_account_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+			),
+		},
```
Comment (tombuildsstuff) [inline azurerm/internal/services/purview/purview_account_resource_test.go:39]:
```suggestion
		},
		data.ImportStep(),
```

File: azurerm/internal/services/purview/purview_account_resource_test.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	})
+}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/purview/purview_account_resource_test.go:41]: we also need a Requires Import test here:

```suggestion
}

func TestAccPurviewAccount_requiresImport(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_purview_account", "test")
	r := PurviewAccountResource{}

	data.ResourceTest(t, r, []resource.TestStep{
		{
			Config: r.basic(data),
			Check: resource.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.RequiresImportErrorStep(r.requiresImport),
	})
}
```

File: azurerm/internal/services/purview/purview_account_resource_test.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+%s
+
+resource "azurerm_purview_account" "test" {
+  name                = "acctestsw%d"
+  resource_group_name = azurerm_resource_group.test.name
+  location            = azurerm_resource_group.test.location
+  sku_name            = "Standard_4"
+}
+`, template, data.RandomInteger)
+}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/purview/purview_account_resource_test.go:72]: we'll need a requires import config here:

```suggestion
}

func (r PurviewAccountResource) requiresImport(data acceptance.TestData) string {
	template := r.basic(data)
	return fmt.Sprintf(`
%s

resource "azurerm_purview_account" "import" {
  name                = azurerm_purview_account.test.name
  resource_group_name = azurerm_purview_account.test.resource_group_name
  location            = azurerm_purview_account.test.location
  sku_name            = azurerm_purview_account.test.sku_name
}
`, template)
}
```

File: website/docs/r/purview_account.html.markdown [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+# azurerm_purview_account
+
+Manages a Purview Account.
+
+## Example Usage
+
+```hcl
+resource "azurerm_resource_group" "example" {
+  name     = "example"
+  location = "West US"
```
Comment (tombuildsstuff) [inline website/docs/r/purview_account.html.markdown:18]: (minor) but could we update this to match the other resources:

```suggestion
  name     = "example-resources"
  location = "West Europe"
```

File: website/docs/r/purview_account.html.markdown [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+---
+
+A `identity` block exports the following:
+
+* `principal_id` - The ID of the Principal (Client) in Azure Active Directory.
+
+* `tenant_id` - The ID of the Azure Active Directory Tenant.
+
+* `type` - The identity type.
```
Comment (tombuildsstuff) [inline website/docs/r/purview_account.html.markdown:73]:
```suggestion
* `type` - The type of Managed Identity assigned to this Purview Account
```

**Review by tombuildsstuff — APPROVED** [review]
hey @allantargino 

Thanks for pushing those changes - taking a look through I've left some comments inline but this should otherwise be good to go - so that we can ship this in today's release I hope you don't mind but I'm going to push a commit to fix these comments 👍 

Thanks!

**Comment by tombuildsstuff** [comment]
Tests pass:

Screenshot 2021-03-11 at 12 41 55

**Comment by allantargino** [comment]
Hi @tombuildsstuff, I always learned so much from reviews like this :)
I highly appreciate it! Thank you so much!

**Pull request merged**